src/scripts/main.js

```diff
@@ -1,3 +1,258 @@
+
+      i++;
+    }
+  }
+
+  setRandom();
+  setRandom();
+  document.addEventListener('keyup', handler);
+}
+
+function hasEmptyTile() {
+  for (let r = 0; r < rows; r++) {
+    for (let c = 0; c < columns; c++) {
+      if (board[r][c] === 0) {
+        return true;
+      }
+    }
+  }
+
+  return false;
+}
+
+function setRandom() {
+  if (!hasEmptyTile()) {
+    return;
+  }
+
+  let isFound = false;
+
+  while (!isFound) {
+    const r = Math.floor(Math.random() * rows);
+    const c = Math.floor(Math.random() * columns);
+
+    if (board[r][c] === 0) {
+      board[r][c] = Math.random() < 0.1 ? 4 : 2;
+
+      const tile = document.getElementById(`${r}-${c}`);
+
+      tile.innerText = '2';
+      tile.classList.add('field-cell--2');
+      isFound = true;
+    }
+  }
+}
+
+function updateTile(tile, num) {
+  tile.innerText = '';
+  tile.classList.value = '';
+  tile.classList.add('field-cell');
+
+  if (num > 0) {
+    tile.innerText = num;
+    tile.classList.add(`field-cell--${num}`);
+  }
+}
+
+const handler = (e) => {
+  switch (e.code) {
+    case 'ArrowLeft': {
+      slideLeft();
+      setRandom();
+
+      break;
+    }
+
+    case 'ArrowRight': {
+      slideRight();
+      setRandom();
+
+      break;
+    }
+
+    case 'ArrowUp': {
+      slideUp();
+      setRandom();
+
+      break;
+    }
+
+    case 'ArrowDown': {
+      slideDown();
+      setRandom();
+
+      break;
+    }
+  }
+
+  lose();
+  win();
+};
+
+function filterZero(row) {
+  return row.filter(num => num !== 0);
+}
+
+function slide(row) {
+  let newRow = filterZero(row);
+
+  for (let i = 0; i < newRow.length; i++) {
+    if (newRow[i] === newRow[i + 1]) {
+      newRow[i] *= 2;
+      newRow[i + 1] = 0;
+      score.textContent = `${+score.textContent + newRow[i]}`;
+    }
+  }
+
+  newRow = filterZero(newRow);
+
+  while (newRow.length < columns) {
+    newRow.push(0);
+  }
+
+  return newRow;
+}
+
+function slideLeft() {
+  for (let r = 0; r < rows; r++) {
+    let row = board[r];
+
+    row = slide(row);
+    board[r] = row;
+
+    for (let c = 0; c < columns; c++) {
+      const tile = document.getElementById(`${r}-${c}`);
+      const num = board[r][c];
+
+      updateTile(tile, num);
+    }
+  }
+}
+
+function slideRight() {
+  for (let r = 0; r < rows; r++) {
+    let row = board[r];
+
+    row.reverse();
+    row = slide(row);
+    row.reverse();
+    board[r] = row;
+
+    for (let c = 0; c < columns; c++) {
+      const tile = document.getElementById(`${r}-${c}`);
+      const num = board[r][c];
+
+      updateTile(tile, num);
+    }
+  }
+}
+
+function slideUp() {
+  for (let c = 0; c < columns; c++) {
+    let row = [board[0][c], board[1][c], board[2][c], board[3][c]];
+
+    row = slide(row);
+
+    for (let r = 0; r < columns; r++) {
+      board[r][c] = row[r];
+
+      const tile = document.getElementById(`${r}-${c}`);
+      const num = board[r][c];
+
+      updateTile(tile, num);
+    }
+  }
+}
+
+function slideDown() {
```

Comment:
there are many similarities in these functions `slideDown` `slideUp`. Merge into one by passing different parameters to avoid duplicate code or leave these two functions, and in the third auxiliary take out everything that is the same in these two